### DISCUSSION TITLE
KubernetesExecutor pod consumes too much momory, how ?

### DISCUSSION THREAD

**Discussion by NickYadance** [apache/airflow#23862] [General]
Airflow in Kubernetes with `KubernetesExecutor` will schedule pod for every operator. In our production environment, each pod is taking up about 140MB memory no matter how simple the operator is. The Grafana metrics views look like this:
<img width="663" alt="image" src="https://user-images.githubusercontent.com/10060849/169777642-eb64ceef-96ab-4bbb-a9d0-912686f97d81.png">

With operator defined as below:
```python
common = BashOperator(task_id='common',
                          bash_command='sleep 600')
```
The `kubectl top` info shows that the pod is taking ~100MB memory. But only 3 processes are running in the pod.
<img width="537" alt="image" src="https://user-images.githubusercontent.com/10060849/169778542-c812977d-95e8-4560-b5a0-cd6400903ebc.png">
<img width="1753" alt="image" src="https://user-images.githubusercontent.com/10060849/169778895-0c846966-4b25-4054-ad9f-32e79fc0fef7.png">
 
`pmap` info for each process. The `airflow task run` command takes much more memory than i expected. But after checking its source code i really don't find anything which should take up so much memory. Any suggestions on this? 
<img width="763" alt="image" src="https://user-images.githubusercontent.com/10060849/169779116-b9914b79-96b5-4e64-80a2-5dfd6c6cfb5f.png">

**Comment by potiuk** [comment]
This is correct. Pod uses image which starts new python interpreter and initializes whole airflow system.  It you want to optimize execution of small tasks - you should use CeleryKubernetesExecutor and send your small tasks to Celery workers. celery workers are much more efficient in handling small tasks because they are using pre-execuged interpreter fotlr many tasks.

See the docs of Airflow for details

**Reply by NickYadance** [reply]
When using `CeleryKubernetesExecutor`, the `RSS` taking by task process is the same as in Kubernetes Pod, about 100MB as `ps -aux` shows
<img width="1654" alt="image" src="https://user-images.githubusercontent.com/10060849/169984016-59b45a65-0e0b-4ec0-8f85-3a9bd4fdcc3a.png">
 

As Celery and Kubernetes are both using `airflow run task --local` to kick off task, when it comes to memory, the only difference between these two should be that **Kubernetes consumes memory in  isolated pod, and Celery does that in a shared pod.** So Celery benefits from pod sharing, but not magic to reduce resource usage for real task process.

Although ~100MB is no big deal for large cluster, but it does seem too heavy for task runner.  Maybe the task runner should be singleton in the scope of vm to reduce resource usage.

**Reply by potiuk** [reply]
What REALLY do you mean as singleton in the scope of vm ? When you use Celery with pre-fork model for workers they are forking from the same interpreter, and forking uses copy-on-write- mechanism, which means that unless shared memory is not changed, ti will not take extra memory. Just a basic fork model - and there always be some memory that stays as shared and some memory that is copied-on-write when you modify it. Is it what you mean by the "singleton" approach ? 

Or you have other ideas how to improve the memory usage (PRs are most welcome).

**Reply by potiuk** [reply]
Maybe you simply do not realize that RSS is a bad metrics when you have shared memory because the shared memory is reported for every process that uses the same non-dirty shared memory.

**Reply by NickYadance** [reply]
I understand. Now i've been  applying `CeleryKubernetesOperator` to my Airflow cluster for quite a while and it works well.

By `singleton`, i was concerned about Airflow task runner, which is kicked off by `airflow task` command, creating duplicate Database connections and other memory-consuming stuffs which wastes resources. However all the tasks are running from `subprocess` or `fork` from Celery worker process. The Celery worker process(if prefork) is the `singleton` one . 

RSS is a bad metric to judge memory usage for a single process. But things get a little complicated when it comes to k8s as 
`container_memory_rss` could cause pod oom-killed. 

I tried setting Celery worker to `gevent` with high concurrency as i have lots of long-running sensor tasks.  And the memory usage of the worker pod reported from `kubectl top po` or prometheus is almost linear to the number of tasks. Adding one task will increase 100MB+ memory metric so i was really worried about pod getting OOM killed.

So my final plan is to send short-running operators like `ShortCurcuitOperator` to Celery, and long-running operators to k8s. And set low memory limit(16G/32G) for worker pod as the tasks finished quickly.

**Reply by potiuk** [reply]
> So my final plan is to send short-running operators like ShortCurcuitOperator to Celery, and long-running operators to k8s. And set low memory limit(16G/32G) for worker pod as the tasks finished quickly.

This is precisely why we have [CeleryKubernetesExecutor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery_kubernetes.html) and you apparently arrived to the very reason why it has been created.

**Reply by potiuk** [reply]
> By singleton, i was concerned about Airflow task runner, which is kicked off by airflow task command, creating duplicate Database connections and other memory-consuming stuffs which wastes resources. However all the tasks are running from subprocess or fork from Celery worker process. The Celery worker process(if prefork) is the singleton one .

Well if you have an idea how to improve stuff - PRs are most welcome. The reason why it is like that is very simple, if you want to run distributed sytem with independent process you have to have a way to communicate with the DB for example. And no matter what some way of communication has to be there. So saying 'singleton' you do not even scratch the surface of what it involves.

FYI. **some** of the DB connection duplication will be handled by introducing https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-44+Airflow+Internal+API which is something we plan to start working on soon. Then the "DB overhead" for communication will be replaced with GRPC /Thrift overhead. It won't go away, might get smaller (or not) - generally speaking communication in distribution system - this way or the other is going to happen. There is no magic to solve it.

**Reply by NickYadance** [reply]
Good one. Thanks for your support.
